File: CIFAR-10_CNN/CIFAR-10_CNN.cpp
#include <fstream>
#include <iostream>
#include <omp.h>
#include <stdio.h>
#include <time.h>

#include "../Neural_Networks.h"

void Read_CIFAR_10(string path, int number_training, int number_test, float **input, float **target_output) {
	ifstream file;

	string filename[] = { "data_batch_1.bin", "data_batch_2.bin", "data_batch_3.bin", "data_batch_4.bin", "data_batch_5.bin", "test_batch.bin" };

	for (int h = 0; h < 6; h++) {
		file.open(path + filename[h]);

		if (!file.is_open()) {
			cerr << "[Read_Data], " + path + filename[h] + " not found" << endl;
			return;
		}
		file.close();
	}
	for (int h = 0, index = 0; h < number_training; h++) {
		unsigned char value;

		if (h % 10000 == 0) {
			if (file.is_open()) {
				file.close();
			}
			file.open(path + filename[index++], ifstream::binary);
		}
		file.read((char*)(&value), 1);

		for (int j = 0; j < 10; j++) {
			target_output[h][j] = (j == value);
		}
		for (int j = 0; j < 3 * 32 * 32; j++) {
			file.read((char*)&value, 1);
			input[h][j] = value / 255.0;
		}
	}
	file.close();
	file.open(path + filename[5], ifstream::binary);

	for (int h = number_training; h < number_training + number_test; h++) {
		unsigned char value;

		file.read((char*)(&value), 1);

		for (int j = 0; j < 10; j++) {
			target_output[h][j] = (j == value);
		}
		for (int j = 0; j < 3 * 32 * 32; j++) {
			file.read((char*)&value, 1);
			input[h][j] = value / 255.0;
		}
	}
	file.close();
}

int main() {
	int batch_size = 100;
	int number_iterations = 100;
	int number_training = 50000;
	int number_test = 10000;

	float **input = new float*[number_training + number_test];
	float **target_output = new float*[number_training + number_test];

	double epsilon = 0.001;
	double decay_rate = 0.977;
	double learning_rate = 0.001;

	string path;

	vector<int> number_nodes = { 3 * 32 * 32, 10 };

	vector<Layer*> layer;

	// train from scratch *****************
	Neural_Networks NN = Neural_Networks();

	layer.push_back(NN.Add(new Layer("CIFAR-10", 3, 32, 32)));
	layer.push_back(NN.Add(new Layer("BN,ReLU", 48, 32, 32)));	// batch normalization, ReLU activation
	layer.push_back(NN.Add(new Layer("BN,ReLU", 48, 32, 32)));	// batch normalization, ReLU activation
	layer.push_back(NN.Add(new Layer("BN,ReLU", 48, 32, 32)));	// batch normalization, ReLU activation
	layer.push_back(NN.Add(new Layer("", 48, 16, 16)));
	layer.push_back(NN.Add(new Layer("BN,ReLU", 48, 16, 16)));	// batch normalization, ReLU activation
	layer.push_back(NN.Add(new Layer("BN,ReLU", 96, 16, 16)));	// batch normalization, ReLU activation
	layer.push_back(NN.Add(new Layer("BN,ReLU", 96, 16, 16)));	// batch normalization, ReLU activation
	layer.push_back(NN.Add(new Layer("BN,ReLU", 96, 16, 16)));	// batch normalization, ReLU activation
	layer.push_back(NN.Add(new Layer("", 96, 8, 8)));
	layer.push_back(NN.Add(new Layer("BN,ReLU", 384)));			// batch normalization, ReLU activation
	layer.push_back(NN.Add(new Layer("CE,softmax", 10)));		// cross-entropy loss, softmax activation

	layer[1]->Connect(layer[0], "W,kernel(3x3)");	 // 3x3 convolution
	layer[2]->Connect(layer[1], "W,kernel(3x3),DS"); // 3x3 depthwise separable convolution
	layer[3]->Connect(layer[2], "W");				 // 1x1 convolution
	layer[4]->Connect(layer[3], "P,max");			 // 2x2 max pooling
	layer[5]->Connect(layer[4], "W,kernel(3x3),DS"); // 3x3 depthwise separable convolution
	layer[6]->Connect(layer[5], "W");				 // 1x1 convolution
	layer[7]->Connect(layer[6], "W,kernel(3x3),DS"); // 3x3 depthwise separable convolution
	layer[8]->Connect(layer[7], "W");				 // 1x1 convolution
	layer[9]->Connect(layer[8], "P,max");			 // 2x2 max pooling
	layer[10]->Connect(layer[9], "W");				 // fully connected
	layer[11]->Connect(layer[10], "W");				 // fully connected

	srand(0); NN.Initialize(0.01);
	// ************************************

	// or load pretrained model
	// Neural_Networks NN = Neural_Networks("CIFAR-10_CNN.txt");

#ifndef Neural_Networks_CUDA_H
	int number_threads;

	cout << "The number of threads : ";
	cin >> number_threads;
	cin.ignore();

	omp_set_num_threads(number_threads);
#endif

	cout << "path where CIFAR-10 dataset is : ";
	getline(cin, path);

	for (int h = 0; h < number_training + number_test; h++) {
		input[h] = new float[number_nodes.front()];
		target_output[h] = new float[number_nodes.back()];
	}
	Read_CIFAR_10(path, number_training, number_test, input, target_output);

	for (int g = 0, time = clock(); g < number_iterations; g++) {
		int score[2] = { 0, };

		float **output = new float*[batch_size];

		double loss = NN.Train(batch_size, number_training, input, target_output, learning_rate, epsilon);

		for (int h = 0; h < batch_size; h++) {
			output[h] = new float[number_nodes.back()];
		}
		for (int i = 0, h = 0; i < number_training + number_test; i++) {
			if (++h == batch_size || i == number_training + number_test - 1) {
				NN.Test(h, &input[i - h + 1], output);

				for (int g = 0, argmax; g < h; g++) {
					double max = 0;

					for (int j = 0; j < number_nodes.back(); j++) {
						if (max < output[g][j]) {
							max = output[g][argmax = j];
						}
					}
					score[(i - h + g + 1 < number_training) ? (0) : (1)] += (int)target_output[i - h + g + 1][argmax];
				}
				h = 0;
			}
		}
		printf("."); NN.Save("NN.txt");

		printf("score: %d / %d, %d / %d  loss: %lf  step %d  %.2lf sec\n", score[0], number_training, score[1], number_test, loss, g + 1, (double)(clock() - time) / CLOCKS_PER_SEC);
		learning_rate *= decay_rate;

		for (int h = 0; h < batch_size; h++) {
			delete[] output[h];
		}
		delete[] output;
	}

	for (int h = 0; h < number_training + number_test; h++) {
		delete[] input[h];
		delete[] target_output[h];
	}
	delete[] input;
	delete[] target_output;

	return 0;
}

File: CIFAR-10_CNN/CIFAR-10_CNN.py
import numpy as np
import time
from Neural_Networks import NNGPU

def Read_CIFAR_10(path, number_training, number_test, image, target_output):
    file = open(path + "data_batch_1.bin", 'rb')
    filename = ["data_batch_1.bin", "data_batch_2.bin", "data_batch_3.bin", "data_batch_4.bin", "data_batch_5.bin", "test_batch.bin"]
    index = 0
    
    for h in range(number_training):
        if h % 10000 == 0:
            if not file.closed:
                file.close()
                
            file = open(path + filename[index], 'rb')
            index += 1

        label = int.from_bytes(file.read(1), byteorder='little')

        for j in range(10):
            target_output[h][j] = (j == label)

        for j in range(3 * 32 * 32):
            image[h][j] = int.from_bytes(file.read(1), byteorder='little') / 255

    file.close()
    file = open(path + filename[5], 'rb')

    for h in range(number_training, number_training + number_test):
        label = int.from_bytes(file.read(1), byteorder='little')

        for j in range(10):
            target_output[h][j] = (j == label)

        for j in range(3 * 32 * 32):
            image[h][j] = int.from_bytes(file.read(1), byteorder='little') / 255
            
    file.close()
    

batch_size      = 100
decay_rate      = 0.977
epsilon         = 0.001
learning_rate   = 0.001
number_iterations = 50
number_training = 50000
number_test     = 10000

# train from scratch
NN = NNGPU.Neural_Networks()

NN.Add_Layer("CIFAR-10", 3, 32, 32)
NN.Add_Layer("BN,ReLU", 48, 32, 32) # batch normalization, ReLU activation
NN.Add_Layer("BN,ReLU", 48, 32, 32) # batch normalization, ReLU activation
NN.Add_Layer("BN,ReLU", 48, 32, 32) # batch normalization, ReLU activation
NN.Add_Layer("", 48, 16, 16)
NN.Add_Layer("BN,ReLU", 48, 16, 16) # batch normalization, ReLU activation
NN.Add_Layer("BN,ReLU", 96, 16, 16) # batch normalization, ReLU activation
NN.Add_Layer("BN,ReLU", 96, 16, 16) # batch normalization, ReLU activation
NN.Add_Layer("BN,ReLU", 96, 16, 16) # batch normalization, ReLU activation
NN.Add_Layer("", 96, 8, 8)
NN.Add_Layer("BN,ReLU", 384)        # batch normalization, ReLU activation
NN.Add_Layer("CE,softmax", 10)      # cross-entropy loss, softmax activation

NN.Connect(1, 0, "W,kernel(3x3)")       # 3x3 convolution
NN.Connect(2, 1, "W,kernel(3x3),DS")    # 3x3 depthwise separable convolution
NN.Connect(3, 2, "W")                   # 1x1 convolution
NN.Connect(4, 3, "P,max")               # 2x2 max pooling
NN.Connect(5, 4, "W,kernel(3x3),DS")    # 3x3 depthwise separable convolution
NN.Connect(6, 5, "W")                   # 1x1 convolution
NN.Connect(7, 6, "W,kernel(3x3),DS")    # 3x3 depthwise separable convolution
NN.Connect(8, 7, "W")                   # 1x1 convolution
NN.Connect(9, 8, "P,max")               # 2x2 max pooling
NN.Connect(10, 9, "W")                  # fully connected
NN.Connect(11, 10, "W")                 # fully connected

NN.Initialize(0, 0.01)


# or load pretrained model
# NN = Neural_Networks("CIFAR-10_CNN.txt")

image = np.zeros((number_training + number_test, 3 * 32 * 32), dtype='f')
target_output = np.zeros((number_training + number_test, 10), dtype='f')

path = input("path where CIFAR-10 dataset is : ")
Read_CIFAR_10(path, number_training, number_test, image, target_output)

start = time.time()

for f in range(number_iterations):
    score = [0, 0]

    loss = NN.Train(batch_size, number_training, image, target_output, learning_rate, epsilon)

    h = 0

    for i in range(number_training + number_test):
        h = h + 1
        
        if h == batch_size or i == number_training + number_test - 1:
            _input = np.array(image[i - h + 1])
            output = np.empty((h, 10), dtype='f')

            for g in range(1, h):
                _input = np.append(_input, image[i - h + g + 1])

            NN.Test(h, _input, output)
            
            for g in range(h):                
                score[0 if i - h + g + 1 < number_training else 1] += int(target_output[i - h + g + 1][np.argmax(output[g])])
                
            h = 0

    print('.', end='')
    NN.Save("NN.txt")
    print('score: {} / {}, {} / {}, loss = {:.6f}, step {}, {:.2f} sec'.format(score[0], number_training, score[1], number_test, loss, f + 1, time.time() - start))
    
    learning_rate *= decay_rate

